core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+        let banking_stage = BankingStage::new_num_threads(
+            block_production_method.clone(),
+            &cluster_info,
+            &poh_recorder,
+            non_vote_receiver,
+            tpu_vote_receiver,
+            gossip_vote_receiver,
+            BankingStage::num_threads(),
+            None,
+            replay_vote_sender,
+            None,
+            connection_cache,
+            bank_forks.clone(),
+            prioritization_fee_cache,
+            false,
+        );
+
+        let (&slot_before_next_leader_slot, &raw_base_event_time) = self
+            .banking_trace_events
+            .freeze_time_by_slot
+            .range(parent_slot..)
+            .next()
+            .expect("timed hashes");
+
+        let base_event_time = raw_base_event_time - warmup_duration;
+        let base_simulation_time = SystemTime::now();
+
+        let sender_thread = thread::Builder::new().name("solSimSender".into()).spawn({
+            let exit = exit.clone();
+
+            move || {
+                let (mut non_vote_count, mut non_vote_tx_count) = (0, 0);
+                let (mut tpu_vote_count, mut tpu_vote_tx_count) = (0, 0);
+                let (mut gossip_vote_count, mut gossip_vote_tx_count) = (0, 0);
+
+                info!("start sending!...");
+                let timed_batches_to_send = self.banking_trace_events.packet_batches_by_time.range(base_event_time..);
+                info!(
+                    "simulating banking trace events: {} out of {}, starting at slot {} (based on {} from traced event slot: {}) (warmup: -{:?})",
+                    timed_batches_to_send.clone().count(),
+                    self.banking_trace_events.packet_batches_by_time.len(),
+                    self.first_simulated_slot,
+                    {
+                        let raw_base_event_time: chrono::DateTime<chrono::Utc> = raw_base_event_time.into();
+                        raw_base_event_time.format("%Y-%m-%d %H:%M:%S.%f")
+                    },
+                    slot_before_next_leader_slot,
+                    warmup_duration,
+                );
+                let mut simulation_duration_since_base = Duration::default();
+                let (
+                    mut last_log_duration,
+                    mut last_tx_count,
+                    mut last_non_vote_count,
+                    mut last_tpu_vote_tx_count,
+                    mut last_gossip_vote_tx_count
+                ) = (Duration::default(), 0, 0, 0, 0);
+                for (&event_time, (label, batches_with_stats)) in timed_batches_to_send {
+                    let required_duration_since_base =
+                        event_time.duration_since(base_event_time).unwrap();
+
+                    // Busy loop for most accurate sending timings
+                    loop {
+                        if simulation_duration_since_base > required_duration_since_base {
+                            break;
+                        }
```

Comment:
seems like a reasonable use for a `while` loop?

Reply:
so true: 6ff38fc92735fa77f0348cc577c330eb3e9a42c5

also, i fixed off-by-one error. strictly speaking, i think it's okay to send batch if `simulation_duration_since_base` *equals to* `required_duration_since_base`.